src/Sylius/Component/Core/StateGuard/CompleteStepGuard.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Core\StateGuard;
+
+use Sylius\Component\Core\Model\OrderInterface;
+
+class CompleteStepGuard
+{
+    /**
+     * When address is required? It seems like always
+     * But should we in any instance block the state change? No
+     */
+    public function address(OrderInterface $order): bool
+    {
+        return false === $order->isEmpty();
+    }
+
+    /**
+     * When shipping address / method is required? When any item requires shipping
+     * But should we in any instance block the state change? No
+     */
+    public function selectShipping(OrderInterface $order): bool
+    {
+        return false === $order->isEmpty();
+    }
+
+    /**
+     * When payment method is required? When order is not free
+     * But should we in any instance block the state change? No
+     */
+    public function selectPayment(OrderInterface $order): bool
+    {
+        return false === $order->isEmpty();
+    }
+
+    public function complete(OrderInterface $order): bool
+    {
+        if ($order->isEmpty()) {
+            return false;
+        }
+
+        if ($order->isShippingRequired() === true) {
+            if ($order->getShippingAddress() === null) {
+                return false;
+            }
+        }
```

Comment:
we could move that to private methods to know better what is going on, it looks like spaghetti made of ifs xD
f.eg if ($this->isShippingReady() && $this->isPaymentReady()...). I think it would improve DX